networkx/algorithms/link_analysis/pagerank_alg.py

```diff
@@ -457,7 +457,15 @@ def _pagerank_scipy(
         return {}
 
     nodelist = list(G)
-    A = nx.to_scipy_sparse_array(G, nodelist=nodelist, weight=weight, dtype=float)
+
+    # Check if the cache exists
+    if "networkx-sparse" in G._cache_it_all:
+        # Get the cached sparse array
+        A = G._cache_it_all["networkx-sparse"]
+    else:
+        A = nx.to_scipy_sparse_array(G, nodelist=nodelist, weight=weight, dtype=float)
+        G._cache_it_all["networkx-sparse"] = A
```

Comment:
Maybe we should put this caching code into the conversion function (I mean `to_scipy_sparse_array` in this example). That saves us maintenance in many places and makes the code for readable in the algorithm itself.

Any downsides to doing that?

Reply:
Yeah I agree, that is probably the best place to add the cacheing!